xclim/utils.py

```diff
@@ -124,6 +124,143 @@ def percentile_doy(arr, window=5, per=.1):
+        if np.all(da.lon > 0) and np.any(lon_bnds < 0):
+            lon_bnds[lon_bnds < 0] += 360
+
+        lon_cond = (da.lon >= lon_bnds.min()) & (da.lon <= lon_bnds.max())
+    else:
+        lon_cond = (da.lon >= da.lon.min()) & (da.lon <= da.lon.max())
+
+    if lat_bnds:
+        lat_bnds = np.asarray(lat_bnds)
+        lat_cond = (da.lat >= lat_bnds.min()) & (da.lat <= lat_bnds.max())
+    else:
+        lat_cond = (da.lat >= da.lat.min()) & (da.lat <= da.lat.max())
+
+    if year_bnds:
+        year_bnds = np.asarray(year_bnds)
+        if len(year_bnds) == 1:
+            time_cond = da.time.dt.year == year_bnds
+        else:
+            time_cond = (da.time.dt.year >= year_bnds.min()) & (da.time.dt.year <= year_bnds.max())
+    else:
+        time_cond = (da.time.dt.year >= da.time.dt.year.min()) & (da.time.dt.year <= da.time.dt.year.max())
+
+    return da.where(lon_cond & lat_cond & time_cond, drop=True)
+
+
+def subset_gridpoint(da, lon, lat, year_bnds=''):
+    """Extract a nearest gridpoint from datarray based on lat lon coordinate.
+    Time series can optionally be subsetted by year(s)
+
+            Return a subsetted data array (or dataset) for the grid point falling nearest the input longitude and latitude
+            coordinates. Optionally subset the data array for years falling within provided year bounds
+
+            Parameters
+            ----------
+            da : xarray.DataArray or xarray.DataSet
+              Input data.
+            lon : float
+              longitude coordinate.
+            lat:  float
+              latitude coordinate.
+            year_bnds (optional) :  list of floats or float
+              List of maximum and minimum year bounds. If single float returns one year only.
+              Defaults to all years in original data-array
+
+
+            Returns
+            -------
+            xarray.DataArray or xarray.DataSet
+              subsetted data array or dataset
+
+        Examples
+        --------
+
+        >>> from xclim import utils
+        >>> ds = xr.open_dataset('pr.day.nc')
+        Subset lat lon point and multiple years
+        >>> prSub = utils.subset_gridpoint(ds.pr, lon=-75,lat=45,year_bnds=[1990,1999])
+        Subset lat, lon point and single year
+        >>> prSub = utils.subset_gridpoint(ds.pr, lon=-75,lat=45,year_bnds=1990) # one year only
+         Subset mutliple variables in a single dataset
+        >>> ds = xr.open_mfdataset(['pr.day.nc','tas.day.nc'])
+        >>> dsSub = utils.subset_gridpoint(ds, lon=-75,lat=45,year_bnds=[1990,1999])
+            """
+
+    # adjust for files with all postive longitudes if necessary
+    if np.all(da.lon > 0) and lon < 0:
+        lon += 360
+
+    try:
+        out = da.sel(lon=lon, lat=lat, method='nearest')
+
+    except:
+
+        dist = np.hypot(da.lat - lat, da.lon - lon)
+        iy, ix = np.unravel_index(np.argmin(dist, axis=None), da.lat.shape)
```

Comment:
This is an approximation as it does not measure distances in km. We could assume the earth is a sphere to convert degrees in km to keep things simple.

Reply:
geopy will provide the functionality as far as I understand. It's mostly plug and play.
https://geopy.readthedocs.io/en/stable/#module-geopy.distance

Reply:
used pyproj to achieve this.  seems easier to calculate distances between point of interest as all gripdpoints without looping